chainer/optimizer.py

```diff
@@ -27,153 +30,313 @@ def exponential_decay_noise(xp, shape, dtype, hook, opt):
     return xp.random.normal(0, std, shape).astype(dtype)
 
 
-class Optimizer(object):
-    """Base class of all numerical optimizers.
+class Hyperparameter(object):
 
-    This class provides basic features for all optimization methods. It
-    optimizes parameters of a *target link*. The target link is registered via
-    the :meth:`setup` method, and then the :meth:`update` method updates its
-    parameters based on a given loss function.
+    """Set of hyperparameters of an optimizer.
 
-    Each optimizer implementation must be defined as a child class of
-    Optimizer. It must override :meth:`update` method. An optimizer can use
-    *internal states* each of which is tied to one of the parameters. State is
-    a dictionary of serializable values (typically arrays of size same as
-    the corresponding parameters). In order to use state dictionaries, the
-    optimizer must override :meth:`init_state` method (or its CPU/GPU versions,
-    :meth:`init_state_cpu` and :meth:`init_state_gpu`).
+    This is a utility class to provide a set of hyperparameters for update
+    rules and an optimizer. It can be used as a usual object.
 
-    If the optimizer is based on single gradient computation (like
-    most first-order methods), then it should inherit :class:`GradientMethod`,
-    which adds some features dedicated for the first order methods.
+    A hyperparameter object can hold a reference to its parent hyperparameter
+    object. When an attribute does not exist in the child hyperparameter, it
+    automatically refers to the parent. We typically set the hyperparameter of
+    the gradient method as the parent of the hyperparameter of each update
+    rule. It enables us to centralize the management of hyperparameters (e.g.
+    we can change the learning rate of all update rules just by modifying the
+    hyperparameter of the central optimizer object), while users can freely
+    customize the hyperparameter of each update rule if needed.
 
-    Optimizer instance also supports *hook functions*. Hook function is
-    registered by the :meth:`add_hook` method. Each hook function is called
-    in registration order in advance of the actual parameter update.
+    Args:
+        parent (Hyperparameter): Parent hyperparameter.
+
+    """
+    def __init__(self, parent=None):
+        self._parent = parent
+
+    def __getattr__(self, name):
+        return getattr(self._parent, name)
```

Comment:
Because `__getattr__` is called only if the access to the attribute of `name` failed. If there is an attribute `name`, this method won't be called. See: https://docs.python.org/3.6/reference/datamodel.html#object.__getattr__